### DISCUSSION TITLE
How `global` a `MeterProvider` really is?

### DISCUSSION THREAD

**Discussion by ramonrrp** [open-telemetry/opentelemetry-go#5988] [General]
Hi all!
searching for some help here! Sorry if this is not the best channel for this :)

According to docs, a [meterProvider](https://opentelemetry.io/docs/concepts/signals/metrics/#meter-provider) is essential to build instruments, and that's clear to me. Right after the initializing example of metric initialization, we have the topic [aquiring a meter](https://opentelemetry.io/docs/languages/go/instrumentation/#acquiring-a-meter) with this object.

Following this implementation in my tests i was note able to properly use `meterProvider` outside of the package where it's implemented (a package created to hold otel build and initialization code): I can acquire a meter with `var meter = otel.Meter("example.io/package/name")`(for example in a `package main`), build an instrument and call `.Add` method. But this metric  were not been sent to my exporter (stdout)

My goal is to keep build and initialization codes in their own package of the application, so that i cann call this and acquire a `meter` in my `package main`. I guess that's a common situation

Is there a way to receive in a package a `meter` builded in another package or something in this line? Is this a nice practice? The doc brings that usually a `MeterProvider` is builded once and lives the time the app lives too. this has to be builded in the same package allways?

**Comment by dmathieu** [comment]
The global meter provider is not specific to one package. It is shared among the entire runtime of your application.
So yes, you can set a global meter provider in one package and reuse it in all the other ones.

If metrics are not exported properly, there must be another issue within your application. Are you able to setup a small reproduction example?

**Comment by ramonrrp** [comment]
Hi and Thanks @dmathieu!

Follow an example showing. Here is a main package trying to acquire and use a meter. Otel is initializated in the `other` package and this `func` is called in beggining of the execution

```go
package main

import (
	"context"
	"localtest/other"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	otm "go.opentelemetry.io/otel/metric"
)

func main() {
	other.Test()

	mainMeter := otel.GetMeterProvider().Meter("localtest/other/myTestMeter")

	mainCounter, _ := mainMeter.Int64Counter(
		"temporary.call",
		otm.WithUnit("{call}"),
		otm.WithDescription("car counter"),
	)
	mainCounter.Add(context.Background(), 1, otm.WithAttributes(
		attribute.String("model", "Camaro"),
		attribute.Bool("IsCar", true),
	))

}
```

```go
package other

import (
	"context"

	"log"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
	otm "go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

func Test() {
	res, err := newResource()
	if err != nil {
		panic(err)
	}
	meterProvider, err := newMeterProvider(res)
	if err != nil {
		panic(err)
	}

	defer func() {
		if err := meterProvider.Shutdown(context.Background()); err != nil {
			log.Println(err)
		}
	}()

	otel.SetMeterProvider(meterProvider)

	Meter := otel.Meter("myTestMeter")

	otherCounter, err := Meter.Int64Counter(
		"temporary.call",
		otm.WithUnit("{call}"),
		otm.WithDescription("car counter"),
	)
	otherCounter.Add(context.Background(), 1, otm.WithAttributes(
		attribute.String("model", "Monza"),
		attribute.Bool("IsCar", true),
	))

}

func newResource() (*resource.Resource, error) {
	return resource.Merge(resource.Default(),
		resource.NewWithAttributes(semconv.SchemaURL,
			semconv.ServiceName("my-service"),
			semconv.ServiceVersion("0.1.0"),
		))
}

func newMeterProvider(res *resource.Resource) (*metric.MeterProvider, error) {
	metricExporter, err := stdoutmetric.New()
	if err != nil {
		return nil, err
	}

	meterProvider := metric.NewMeterProvider(
		metric.WithResource(res),
		metric.WithReader(metric.NewPeriodicReader(metricExporter,
			// Default is 1m. Set to 3s for demonstrative purposes.
			metric.WithInterval(3*time.Second))),
	)
	return meterProvider, nil
}
```

Here the output that show only the counter generated in the same package where otel is initializated:

```
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"my-service"}},{"Key":"service.version","Value":{"Type":"STRING","Value":"0.1.0"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.32.0"}}],"ScopeMetrics":[{"Scope":{"Name":"myTestMeter","Version":"","SchemaURL":"","Attributes":null},"Metrics":[{"Name":"temporary.call","Description":"car counter","Unit":"{call}","Data":{"DataPoints":[{"Attributes":[{"Key":"IsCar","Value":{"Type":"BOOL","Value":true}},{"Key":"model","Value":{"Type":"STRING","Value":"Monza"}}],"StartTime":"2024-11-20T07:40:23.831377979-03:00","Time":"2024-11-20T07:40:23.831517987-03:00","Value":1}],"Temporality":"CumulativeTemporality","IsMonotonic":true}}]}]}
```

**Reply by dmathieu** [reply]
```golang
defer func() {
		if err := meterProvider.Shutdown(context.Background()); err != nil {
			log.Println(err)
		}
	}()
```

You're shutting down this meter provider at the end of `other.Test()`, so in your main method, the meter provider won't log anything because it has been shut down already.
You need to shutdown only at the end of your entire program.

See this example for a way to do this: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/examples/otel-collector/main.go#L89

**Comment by ramonrrp** [comment]
Hey @dmathieu i was able to reproduce this example and everything worked well!
Thanks a lot for this help! ❤️
